### ISSUE TITLE
smooth versions

### ISSUE THREAD

**Issue by ghorn** [ghorn/atmosphere#7]
It would be great to have something with only `Num/Fractional/Floating` constraints. Maybe smoothed piecewise approximations using sigmoids.

**Comment by dmcclean** [comment]
The guys at Galois were requesting this too, because the `Ord` constraint stops them from using their Ivory C-generation EDSL to compile the `ad` derivative of it, which is part of their vehicle state estimator update equation.

**Comment by ghorn** [comment]
I also need the Ord constraint removed for AD.

**Comment by dmcclean** [comment]
I was under the impression that AD itself (at least from the `ad` package) worked in the presence of an `Ord` constraint but with some potential weirdness at discontinuities?

**Comment by ghorn** [comment]
You may be right. I use my own ad code, which doesn't handle the `Ord` typeclass. I was thinking of providing a smooth approximating function in addition to the existing functions. I have to sit down and plot the thing and see what makes sense.

**Comment by dmcclean** [comment]
Interesting. I'm sure it's at least possible, because I remember seeing a presentation about how it works, I'm not 100% sure if Edward's library does it or not but looking at the haddocks I think it does.

I got partway trying to play around with it in R, but I need to go home and switch computers. Attached is a CSV file with heights in meters, pressures in pascals, and temperatures in kelvin, to save myself repeating that step.

[pressures.txt](https://github.com/ghorn/atmosphere/files/43316/pressures.txt)

**Comment by dmcclean** [comment]
The incredibly naive just-fit-a-20th-order-polynomial approach works reasonably well, maximum residuals around 1K:

![polynomial-fit-t](https://cloud.githubusercontent.com/assets/603347/11403181/8770b01e-9369-11e5-92ba-0b2e043fbb09.png)

Normally I try Floater-Hormann rational fits for this kind of thing because they usually work out better, but I am having trouble persuading R to do it for me and the C# library I usually use is a pain to get started with.

Obviously doing the pressure is easier since it's monotonic and the model is already smooth even though the definition is still technically done piecewise.

**Comment by ghorn** [comment]
Which curve are you fitting? I suggest fitting (density, pressure, temperature) and then deriving the rest from those, because that is how the model is implemented: 

https://github.com/ghorn/atmosphere/blob/master/src/Atmosphere.hs#L115-L149
https://github.com/ghorn/atmosphere/blob/master/src/Atmosphere.hs#L21-L51

**Comment by dmcclean** [comment]
I was fitting temperature since it seems to be the most difficult one. But I wasn't fitting the version that's normalized to sea level.

**Comment by dmcclean** [comment]
I had trouble producing a polynomial approximation for all three components that was accurate enough and not ridiculous.

But now it occurs to me that perhaps there is a more direct approach to getting rid of the `Ord` constraint. Can't we rephrase the whole model in terms of indicator functions for which layer you are in, and use `signum` (from `Num`, not `Ord`) to define the indicator functions?

As far as I can see, that gets you the exact function in the standard atmosphere (except possibly at the exact layer transition altitudes?) in a form that is handled by `ad`? It's also conveniently a form that is handled by Ivory.

I will take a swing at this over the weekend.

**Comment by ghorn** [comment]
You can definitely use signum here but your resulting function wouldn't be differentiable. If you used a quasi-newton method to optimize over this you would lose convergence properties. There's a similar problem with abs.

You could use sigmoids to the same effect though. I have this code lying around:
```haskell
-- | A convenient parameterization of an exponential sigmoid.
data Sigmoid a =
 Sigmoid
 { sigInputCenter :: a
 , sigOutputCenter :: a
 , sigMagnitude :: a
 , sigGain :: a
 } deriving (Functor, Generic, Generic1)

-- | Evaluate a sigmoid function.
runSigmoid :: Floating a => Sigmoid a -> a -> a
runSigmoid s x = mag * 2 * (1/(1 + exp (- 2 * gain * dx)) - 0.5) + offset
  where
    offset = sigOutputCenter s
    mag = sigMagnitude s
    dx = x - sigInputCenter s
    gain = sigGain s

-- | Uses runSigmoid to smoothly saturate a function between given bounds.
-- Function is approximately linear when far from the bounds.
smoothSaturate :: Floating a => (a, a) -> a -> a
smoothSaturate (lb, ub) = runSigmoid sig
  where
    mean = 0.5*(lb + ub)
    magnitude = 0.5*(ub - lb)
    sig =
      Sigmoid
      { sigInputCenter = mean
      , sigOutputCenter = mean
      , sigMagnitude = magnitude
      , sigGain = 1 / magnitude
      }
```

**Comment by dmcclean** [comment]
I'm not totally convinced of that in the `ad` setting. At the transition points the `H(0) = 1/2` and the product rule combine to make the derivative half the lower layer's derivative plus half the higher layer's derivative, which seems simple enough and so the derivative is defined and computable everywhere. Ideally the lower layer and the higher layer models would agree on the value at the transition point, but I don't think it works out that way with approximate arithmetic.

It seems to work out. I made a test implementation with only a `Floating` constraint that agrees with the actual values of `atmosphere` and the values of `diff . (\(a,b,c) -> a) . atmosphere` to within 3 ulps everywhere I looked (a few random spots in the middle of layers, and exhaustively for 20 floats on either side of each layer boundary).

The inverses taken with `Numeric.AD.Newton.inverse` seem to work even when the initial guesses are well on the wrong side of layer boundaries. (Obviously subject to caveats for the temperature one, since the model is non-monotonic.)

Can you say where else I should look for differentiability/convergence problems?

It's a mess but I'll clean it up and run it through the test suite.

**Comment by ghorn** [comment]
I see how you can make the derivative exist at the transition point, but I want the derivative to be sufficiently smooth so that I can optimize over it. Specifically for using Newton's method, I want the function to be optimized over to have continuous first and second derivatives.

For example, absolute value is continuous but the derivative is not. If you use newton's method to minimize `abs(x)` starting with `x = 1`, the first iteration will decide that the minimum is at negative infinity because that is true of the linearization of `abs` evaluated at `1`.

```
>> Prelude Numeric.AD.Newton> take 10 $ extremum abs 1
[1.0,-Infinity,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN]
```

I think you simply got lucky that `inverse` worked. I think after 1 or a few steps, it luckily found itself in the correct region and then quickly converged, but there is no guarantee of robustness.

If you have a working implementation using `signum`, we could put it in with a big warning that it is not 
"smooth", rather it circumvents `Ord`. Then we could also provide a smooth version using my sigmoidal functions, which should be a drop-in replacement for your indicator functions.

Do you have this work on a branch I could look at?

**Comment by dmcclean** [comment]
Right, that makes sense in that case. If I were going to go looking for problems like that I would look where a layer with constant temperature meets up with one that isn't. That temperature function is just legitimately tough to search with non-monotonicity and long flat stretches. The non-temperature ones should have smooth transitions that I think work even for this purpose.

Implementation is at https://github.com/dmcclean/atmosphere/tree/layer-indicators. Note that it doesn't expose precisely the same API, I would finish cleaning that up before submitting a PR for it. (I also pushed dimensional further into the implementation to guide myself on doing it, I can undo all that too if needed, now that I grok it better.)

**Comment by ghorn** [comment]
Feel free to leave dimensional deep in the implementation - it can't hurt. Is it difficult to have the `Ord` and `signum` implementations in side-by-side? I'm not sure if the `Ord` one has any advantages over `signum` at this point.

**Comment by dmcclean** [comment]
We could have them side by side easily.

I haven't done any profiling but I strongly suspect the `Ord` one is faster, since it selects a layer and then does the calculation for that layer. The one without `Ord` has to calculate all layers only to multiply most of them by `0` and sum the results. It's a constant factor of 8 (ignoring however much work the layer search takes).
